### PR TITLE
Refactor the Slack, Email, MS Teams and Attachments

### PR DESCRIPTION
**Pull request by ateska**
# What is done:
- [x] Refactoring of Slack including a support for markdown in Slack messages
- [x] Refactoring of Email
- [x] Generalised work with attachment into a dedicated service `AttachmentRenderingService`, now used by Slack and Email
- [x] General refactoring
- [x] Add Markdown -> Text conversion into `MarkdownToHTMLService`
- [x] Case insensitive a nested dicts for Jinja2

# What needs to be finished

- [ ] Testing
- [ ] Loading jinja2 variables from json file

### CODE REVIEW

**Review by PremyslCerny — APPROVED** [review]

**Comment by ateska** [comment]
Please don't merge this one, I have a refactoring MR on the way.

**Comment by mithunbharadwaj** [comment]
@ateska I added the unit test, If you are still refactoring they might be no longer valid. Perhaps I can remove them and create unit test in a separate MR.

**Comment by mithunbharadwaj** [comment]
@ateska 

1. I found one problem at line 79

Path error is incorrectly raised
    `raise PathError(path=template)`
    
correct way is
	raise PathError(use_case='Slack', invalid_path=template)

Also, this check here is redundant

https://github.com/TeskaLabs/asab-iris/blob/cd961137e1afe92d1e4f710e7f86cc69de531898/asabiris/orchestration/sendslack.py#L112 


```

		if not template.startswith("/Templates/Slack/"):
			raise PathError(path=template)

```

 2. The path validation performed here seems to be repetitive, Given that this kind of check is conducted multiple times, it might be more efficient to encapsulate it within a separate method.

**Comment by mithunbharadwaj** [comment]
@ateska This hard-coded path can be made constant

**/Templates/Slack/** in

```
if not template.startswith("/Templates/Slack/"):
			raise PathError(path=template)

```

# Constant 
TEMPLATE_PREFIX = "/Templates/Slack/"

**Comment by mithunbharadwaj** [comment]
@ateska Shouldn't you be calling **send_files**?


https://github.com/TeskaLabs/asab-iris/blob/cd961137e1afe92d1e4f710e7f86cc69de531898/asabiris/orchestration/sendslack.py#L102

**Comment by ateska** [comment]
> @ateska This hard-coded path can be made constant
> 
> **/Templates/Slack/** in
> 
> ```
> if not template.startswith("/Templates/Slack/"):
> 			raise PathError(path=template)
> ```
> 
> # Constant
> TEMPLATE_PREFIX = "/Templates/Slack/"

Please don't do these constants, when they are used only on one or two places, they just increase the cognitive load with no apparent benefit (to me at least).

File: asabiris/orchestration/sendemail.py [hunk 1]
```diff
@@ -6,206 +6,164 @@
...
+			return body, subject
+
+		else:
+			raise FormatError(format=ext)
+
+
+	def _generate_error_message(self, specific_error: str) -> Tuple[str, str]:
+		try:
+			timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+		except Exception as date_err:
```
Comment (eliska-n) [inline asabiris/orchestration/sendemail.py:130]: what could possibly go wrong in `datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")` ?

File: asabiris/orchestration/sendemail.py [hunk 2]
```diff
@@ -6,206 +6,164 @@
...
+
+		try:
+			error_message = (
+				"Hello!<br><br>"
+				"We encountered an issue while processing your request: <b>{}</b><br>"
+				"Please review your input and try again.<br><br>"
+				"Timestamp: {}<br></pre>"  # Added a <br> here for a new line in HTML
+				"<br>Best regards,<br>Your Team"
+			).format(specific_error, timestamp)
+		except Exception as format_err:
```
Comment (eliska-n) [inline asabiris/orchestration/sendemail.py:142]: again... what can be so wrong about formatting?

File: asabiris/orchestration/sendmsteams.py [hunk 3]
```diff
@@ -45,14 +45,13 @@ async def send_to_msteams(self, msg):
...
-
 		body = msg['body']
-		if not body['template'].startswith("/Templates/MSTeams/"):
-			raise PathError(use_case='MSTeams', invalid_path=body['template'])
+		template = body["template"]
 
-		body["params"] = body.get("params", {})
-		output = await self.JinjaService.format(body["template"], body["params"])
+		if not template.startswith("/Templates/MSTeams/"):
+			raise PathError(use_case='MSTeams', invalid_path=template)
```
Comment (eliska-n) [inline asabiris/orchestration/sendmsteams.py:52]: I guess these errors shoudl be caucht specifically, not just using except Exception

File: asabiris/orchestration/sendslack.py [hunk 4]
```diff
@@ -36,61 +37,57 @@ async def send_to_slack(self, msg):
...
 		# templates must be stores in /Templates/Slack
-		if not body['template'].startswith("/Templates/Slack/"):
-			raise PathError(use_case='Slack', invalid_path=body['template'])
-
-		atts = []
-
-		for a in attachments:
-			# If there are 'template' we render 'template's' else we throw a sweet warning.
-			# If content-type is application/octet-stream we assume there is additional attachments in request else we raise bad-request error.
-			template = a.get('template', None)
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:54]: fallback to `None` is implicit

**Review by eliska-n — CHANGES_REQUESTED** [review]

File: asabiris/orchestration/sendmsteams.py (hunk 3 above)
Comment (ateska) [reply asabiris/orchestration/sendmsteams.py:52]: Not sure what do you mean,

**Comment by mithunbharadwaj** [comment]
Obtained testing instructions:

```
1. Sending emails from template.txt
2. Sending emails from template.md
3. Sending emails from template.html
4. Sending emails with attachments (more than one) in Base64
5. Sending emails with attachments (more than one) as a template.md, format PDF
6. Sending emails with attachments (more than one) as a template.md, format MD
7. Sending emails with attachments (more than one) as a template.md, format HTML
8. Sending emails with attachments (more than one) as a template.html, format PDF
9. Sending emails with attachments (more than one) as a template.html, format MD
10. Sending emails with attachments (more than one) as a template.html, format HTML
11. Sending emails with attachments (more than one) as a template.txt, format TXT
12. Sending slack message from template.txt
13. Sending slack message from template.md
14. Sending slack message from template.html
15. Sending slack message with attachments (more than one) in Base64
16. Sending slack message with attachments (more than one) as a template.md, format PDF
17. Sending slack message with attachments (more than one) as a template.md, format MD
18. Sending slack message with attachments (more than one) as a template.md, format HTML (doesn't look nice in SLack)
19. Sending slack message with attachments (more than one) as a template.html, format PDF
20. Sending slack message with attachments (more than one) as a template.html, format MD
21. Sending slack message with attachments (more than one) as a template.html, format HTML (doesn't look nice in SLack)
22. Sending slack message with attachments (more than one) as a template.txt, format TXT
23. Sending ms teams message from template.txt
24. Sending ms teams message from template.md
25. Sending ms teams message from template.html
26. Sending ms teams message with attachments (more than one) in Base64
27. Sending ms teams message with attachments (more than one) as a template.md, format PDF
28. Sending ms teams message with attachments (more than one) as a template.md, format MD
29. Sending ms teams message with attachments (more than one) as a template.md, format HTML
30. Sending ms teams message with attachments (more than one) as a template.html, format PDF
31. Sending ms teams message with attachments (more than one) as a template.html, format MD
32. Sending ms teams message with attachments (more than one) as a template.html, format HTML
33. Sending ms teams message with attachments (more than one) as a template.txt, format TXT

```
template.txt ,  template.md and  template.html do not exist in 'library/Templates'

**Comment by mithunbharadwaj** [comment]
I am currently testing the email use case:

While sending email :

```
{
    "to": ["Shivashankar <Firstname.Sirname@teskalabs.com>"],
    "cc": ["<mithunshivashankar@gmail.com>","mithuns.shivashankar@teskalabs.com"],
    "body": {
        "template": "/Templates/Email/message.html",
         "params":{
            "name":"Iris"
      }
    }
}
```


I get a response as 
```
{
    "result": "OK"
}
```

But I do not see the email in my account.

**Comment by mithunbharadwaj** [comment]
```
========================================================================= short test summary info ==========================================================================
FAILED test_failsafe_mechanism.py::TestRenderMethod::test_general_exception - Exception: General Exception
FAILED test_failsafe_mechanism.py::TestRenderMethod::test_jinja_template_not_found - TypeError: TemplateNotFound.__init__() missing 1 required positional argument: 'name'
FAILED test_failsafe_mechanism.py::TestRenderMethod::test_jinja_template_syntax_error - TypeError: TemplateSyntaxError.__init__() missing 2 required positional arguments: 'message' and 'lineno'
FAILED test_failsafe_mechanism.py::TestRenderMethod::test_jinja_undefined_error - jinja2.exceptions.UndefinedError: None
FAILED test_failsafe_mechanism.py::TestRenderMethod::test_render_jinja_exception - jinja2.exceptions.TemplateError: Mocked Jinja Error

```


Unit test  test_failsafe_mechanism.py is broken. Must be fixed.

**Comment by mithunbharadwaj** [comment]
```
========================================================================= short test summary info ==========================================================================
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_empty_dict - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_01 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_02 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_03 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_04 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_05 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
FAILED test_nested_dictwith_dots.py::TestNestedDictWithDos::test_nested_dict_06 - AttributeError: module 'asabiris.utils' has no attribute 'create_nested_dict_from_dots_in_keys'
====================================================================== 7 failed, 2 warnings in 0.89s =======================================================================
```

Also Unit test_nested_dictwith_dots is broken. Must be fixed too.

**Comment by mithunbharadwaj** [comment]
```
========================================================================= short test summary info ==========================================================================
FAILED test_sendslack_orchestrator.py::TestSendSlackOrchestrator::test_attachment_rendering - AttributeError: <asabiris.orchestration.sendslack.SendSlackOrchestrator object at 0x109613cd0> does not have the attribute 'render'
FAILED test_sendslack_orchestrator.py::TestSendSlackOrchestrator::test_attachment_rendering_failure - AttributeError: <asabiris.orchestration.sendslack.SendSlackOrchestrator object at 0x1096c3cd0> does not have the attribute 'render'
FAILED test_sendslack_orchestrator.py::TestSendSlackOrchestrator::test_send_to_slack_with_invalid_attachment_path - AttributeError: 'NoneType' object has no attribute 'unformat'
================================================================ 3 failed, 11 passed, 16 warnings in 1.06s =================================================================
```

Also test_sendslack_orchestrator.py is broken. Must be updated

**Comment by mithunbharadwaj** [comment]
Identified a breaking change: Additional Slack API scopes (groups:read, mpim:read, im:read) are now required, beyond the existing chat:write, files: write, files:read, channels:read. README needs updating to reflect these new permissions for accurate setup and testing

```
{
    "result": "ERROR",
    "message": "The request to the Slack API failed. (url: https://www.slack.com/api/conversations.list)\nThe server responded with: {'ok': False, 'error': 'missing_scope', 'needed': 'channels:read,groups:read,mpim:read,im:read', 'provided': 'incoming-webhook,chat:write,files:write,reactions:write,users:write,chat:write.public,files:read,links:write'}",
    "uuid": "89ed9d26-b5e4-4fa5-b1f4-32d65a7dd1fa"
}

```

**Comment by mithunbharadwaj** [comment]
Your approach to the test case seems reasonable given the lack of specific details. Here's how you can structure your assumption for clarity and future reference:

---

**Assumption for Test Case: Sending emails with attachments (more than one) as a template.md, format PDF**

Given the test case: "Sending emails with attachments (more than one) as a template.md, format PDF," and the absence of specific instructions regarding the templates for the email body and attachments, I am proceeding with the following assumptions:

1. **Email Body Template:**
   - I will use `message.md` as the template for the email body. This choice is based on the typical use of `message.md` for email content in similar scenarios.

2. **Attachments:**
   - For the attachments, I will utilize `attachment.md` as the template.
   - Each `attachment.md` will be converted to PDF format, as required by the test case.

**Comment by mithunbharadwaj** [comment]
I observed inconsistent behavior in /send_mail endpoint: When sending two attachments with the format: "HTML" that uses the same template stored in /Templates/Attachment/attachment.md, only one attachment is received. 

However, changing the format to PDF results in both attachments being sent successfully. This suggests a potential issue with handling multiple HTML attachments.

**Pull request merged**